File: 30_day_challenge_2020_December/98_validate_binary_search_tree.py
################################################################
# Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
# Date created      : 20201216
# Problem link      : https://leetcode.com/problems/validate-binary-search-tree/
################################################################

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import math

class Solution:
    # def isValidBST(self, root: TreeNode) -> bool:
    #     def dfs(node, lower, upper):
    #         if not node:
    #             return True
    #         if lower < node.val < upper:
    #             return dfs(node.left, lower, node.val) and dfs(node.right, node.val, upper)
    #         return False
    #     return dfs(root, -math.inf, math.inf)
    
    def isValidBST(self, root: TreeNode, lower = -math.inf, upper = math.inf) -> bool:
        if not root:
            return True
        if lower < root.val < upper:
            return self.isValidBST(root.left, lower, root.val) and self.isValidBST(root.right, root.val, upper)
        return False

File: 30_day_challenge_2020_December/98_validate_binary_search_tree.cpp
/****************************************************************
* Author            : yiorgosynkl (find me in Github: https://github.com/yiorgosynkl)
* Date created      : 20201216
* Problem link      : https://leetcode.com/problems/validate-binary-search-tree/
****************************************************************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    // without INT_MIN and INT_MAX
    bool isValidBST(TreeNode* root) {
        TreeNode* prev = NULL;
        return validate(root, prev);
    }
    bool validate(TreeNode* node, TreeNode* &prev) {
        if (node == NULL) return true;
        if (!validate(node->left, prev)) return false;
        if (prev != NULL && prev->val >= node->val) return false;
        prev = node;
        return validate(node->right, prev);
    }
};
